Clean up transactional behavior in User API

andy-armstrong opened the pull request:
I've removed the ``@transaction.commit_on_success`` decorators from the user preference APIs and instead explicitly used a transaction in the PATCH method that updates multiple user preferences at once. I've written a test that verifies that the rollback behavior works as desired.

While I was here, I also did a sweep for developer messages that were being localized. I've fixed the one place that was doing this, and fixed another message that wasn't unicode.

@cahrens @nasthagiri please review when you have a chance.

cahrens commented on openedx/core/djangoapps/user_api/preferences/api.py:
```diff
@@ -353,11 +350,15 @@ def validate_user_preference_serializer(serializer, preference_key, preference_v
         raise PreferenceValidationError({
-            preference_key: {"developer_message": message, "user_message": message}
+            preference_key: {
+                "developer_message": u"Preference '{preference_key}' cannot be set to an empty value.".format(
```
Note that you can use noop to avoid having to duplicate the string.

(http://blog.bessas.me/post/65775299341/using-gettext-in-django

cahrens commented on openedx/core/djangoapps/user_api/preferences/tests/test_views.py:
```diff
@@ -288,6 +292,56 @@ def test_update_preferences_other_user(self, api_client, user):
+        set_user_preference(self.user, "c", "3")
+
+        # Patch the serializer so that it throws an exception on the second save
+        # serializer_save = patch('openedx.core.djangoapps.user_api.serializers.RawUserPreferenceSerializer.save')
```
Why the commented out code?

andy-armstrong replied:
That was my failed attempt to have one of the updates patched but not another. I eventually settled on patching delete instead which has the same effect. I've removed the bogus comment (and fixed the name of the patch variable).

cahrens commented on openedx/core/djangoapps/user_api/preferences/tests/test_views.py:
```diff
@@ -288,6 +292,56 @@ def test_update_preferences_other_user(self, api_client, user):
+        # serializer_save.start()
+        # self.addCleanup(serializer_save.stop)
+        # serializer_save.side_effect = [Exception, None]
+        delete_serializer.side_effect = [Exception, None]
```
So this works because the 2nd item (b set to None) calls delete? The comments could use some updating.

andy-armstrong replied:
Indeed. I've cleaned up the comments.

cahrens commented:
You forgot about the accounts view PATCH method.

andy-armstrong commented:
@cahrens I can't believe I forgot the accounts view PATCH method. I've implemented it properly now and added a rollback test. I've also addressed your other feedback.

cahrens commented on openedx/core/djangoapps/user_api/accounts/tests/test_views.py:
```diff
@@ -509,3 +510,37 @@ def test_patch_serializer_save_fails(self, serializer_save):
+        """
+        Verify that updating preferences is transactional when a failure happens.
+        """
+        # Send a PATCH request with updates to both the user and the profile models.
```
Name and gender are both profile fields. Did you verify that this test fails without the transaction logic?

I don't think there are any user model fields that can be directly updated through the accounts patch method (email can be updated, but it is multi-step).

andy-armstrong replied:
Good catch. I forgot to test that this one fails first (I did for the other one), and it doesn't as you say. I've tried several variations on patching but have been unable to turn this into a failing test. For some reason when I patch methods in the student Django app, they are not mocked when the code is actually reached. I'm out of time for this tonight so I'll try again tomorrow.

cahrens replied:
It probably has to do with how do_email_change_request is imported in api.py. Try changing the code to just import student.views, then reference do_email_change_request as views.do_email_change_request. I've had similar issues in the past.

cahrens commented on openedx/core/djangoapps/user_api/preferences/tests/test_views.py:
```diff
@@ -288,6 +292,52 @@ def test_update_preferences_other_user(self, api_client, user):
+        # after one of the updates has happened, in which case the whole operation
+        # should be rolled back.
+        delete_user_preference.side_effect = [Exception, None]
+        self.client.login(username=self.user.username, password=self.test_password)
```
Did you verify that this test fails without the transaction logic?

cahrens replied:
OK, I verified this one does fail. But the accounts one does not.

cahrens commented:
LGTM, just want to make sure the tests are actually covering the rollback behavior.

cahrens commented:
:+1:

cahrens commented:
@andy-armstrong please look this one again, then if it looks good, merge once tests pass.
